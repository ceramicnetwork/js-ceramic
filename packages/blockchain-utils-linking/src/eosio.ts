import { AuthProvider } from "./auth-provider";
import { AccountID } from "caip";
import { getConsentMessage, LinkProof } from "./util";
import { normalizeAccountId } from "./ethereum";
import * as sha256Stable from '@stablelib/sha256'
import * as uint8arrays from 'uint8arrays'

const maxWordLength = 12;

export class EosioAuthProvider implements AuthProvider {
  readonly isAuthProvider = true;

  constructor(
      private readonly provider: any,
      private readonly address: string
  ) {}

  async accountId(): Promise<AccountID> {
    const chainId = toCAIPChainId(await this.provider.getChainId());
    return new AccountID({
      address: this.address,
      chainId: `eosio:${chainId}`,
    });
  }

  async authenticate(message: string): Promise<string> {
    const accountID = await this.accountId()
    const signedPayload = await toSignedPayload(message, accountID, this.provider)
    const signatureBytes = uint8arrays.fromString(signedPayload)
    const digest = sha256Stable.hash(signatureBytes)
    return `0x${uint8arrays.toString(digest, 'base16')}`
  }

  async createLink(did: string): Promise<LinkProof> {
    const consentMessage = getConsentMessage(did);
    const accountID = await this.accountId();
    const signedPayload = await toSignedPayload(
        consentMessage.message,
        accountID,
        this.provider
    );
    return {
      version: 1,
      type: "eosio",
      message: consentMessage.message,
      signature: signedPayload,
      account: accountID.toString(),
      timestamp: consentMessage.timestamp,
    };
  }

  withAddress(address: string): AuthProvider {
    return new EosioAuthProvider(this.provider, address);
  }
}

function toCAIPChainId(chainId: string): string {
  return chainId.substr(0, 32);
}

function sanitize(str: string, size: number): string {
  return str
      .replace(/\s/g, " ")
      .replace(new RegExp(`(\\S{${size}})`, "g"), "$1 ");
}

export function toPayload(message: string, accountID: AccountID): string {
  const { address, chainId } = accountID;
  const payload = `${message} [For: ${address} on chain: ${chainId}]`;
  return sanitize(payload, maxWordLength);
}

export async function toSignedPayload(
    message: string,
    accountID: AccountID,
    provider: any
): Promise<string> {
  accountID = normalizeAccountId(accountID);
  const {
    chainId: { reference: requestedChainId },
    address,
  } = accountID;
  const accountName = await provider.getAccountName();
  const chainId = toCAIPChainId(await provider.getChainId());

  if (chainId !== requestedChainId) {
    throw new Error(
        `Provider returned a different chainId than requested [returned: ${chainId}, requested: ${requestedChainId}]`
    );
  }
  if (accountName !== address) {
    throw new Error(
        `Provider returned a different account than requested [returned: ${accountName}, requested: ${address}]`
    );
  }
  const payload = toPayload(message, accountID);
  const [key] = await provider.getKeys();
  return provider.signArbitrary(key, payload);
}
