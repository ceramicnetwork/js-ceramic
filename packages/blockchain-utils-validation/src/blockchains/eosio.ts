import { AccountID } from "caip";
import { SigningTools } from "@smontero/eosio-signing-tools";
import { BlockchainHandler } from "../blockchain-handler";
import * as linking from "@ceramicnetwork/blockchain-utils-linking";

const namespace = "eosio";

export async function validateLink(
  proof: linking.LinkProof
): Promise<linking.LinkProof | null> {
  const { message, signature, account } = proof;
  const accountID = new AccountID(account);
  const { address, chainId } = accountID;
  try {
    const success = await SigningTools.verifySignature({
      chainId: chainId.reference,
      account: address,
      signature,
      data: linking.eosio.toPayload(message, accountID),
    });
    return success ? proof : null;
  } catch (error) {
    console.warn(error);
    return null;
  }
}

const Handler: BlockchainHandler = {
  namespace,
  validateLink,
};

export default Handler;
