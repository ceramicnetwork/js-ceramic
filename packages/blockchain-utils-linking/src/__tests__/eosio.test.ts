import { EOSIOProvider } from "@smontero/eosio-local-provider";
import { EosioAuthProvider } from "../eosio";

const did = "did:3:bafysdfwefwe";
const telosTestnetChainId =
  "1eaa0824707c8c16bd25145493bf062aecddfeb56c736f6ba6397f3195f33c9f";
const jungleChainId =
  "2a02a0053e5a8cf73a56ba0fda11e4d92e0238a4a2aa74fccf46d5a910746840";
const telosTestnetAccount = "testuser1111";
const jungleAccount = "idx3idctest1";
const telosTestnetProvider = new EOSIOProvider({
  chainId: telosTestnetChainId,
  account: telosTestnetAccount,
  keys: {
    EOS6uUc8fYoCdyz7TUAXqHvRbU7QnVirFuvcAW6NMQqBabdME6FnL:
      "5KFFFvKioakMpt8zWnyGKnLaDzzUSqy5V33PHHoxEam47pLJmo2",
  },
});
const jungleProvider = new EOSIOProvider({
  chainId: jungleChainId,
  account: jungleAccount,
  keys: {
    EOS7f7hdusWKXY1cymDLvUL3m6rTLKmdyPi4e6kquSnmfVxxEwVcC:
      "5JRzDcbMqvTJxjHeP8vZqZbU9PwvaaTsoQhoVTAs3xBVSZaPB9U",
  },
});

beforeAll(() => {
  global.Date.now = jest.fn().mockImplementation(() => 666000);
});

afterAll(() => {
  jest.clearAllMocks();
});

test("accountId", async () => {
  const authProvider = new EosioAuthProvider(
    telosTestnetProvider,
    telosTestnetAccount
  );
  await expect(authProvider.accountId()).resolves.toMatchSnapshot();
});

describe("createLink", () => {
  test("generate proof on telos testnet", async () => {
    const authProvider = new EosioAuthProvider(
      telosTestnetProvider,
      telosTestnetAccount
    );
    await expect(authProvider.createLink(did)).resolves.toMatchSnapshot();
  });

  test("generate proof on jungle testnet", async () => {
    const authProvider = new EosioAuthProvider(jungleProvider, jungleAccount);
    await expect(authProvider.createLink(did)).resolves.toMatchSnapshot();
  });

  test("fail on telos testnet account for jungle provider", async () => {
    const authProvider = new EosioAuthProvider(
      jungleProvider,
      telosTestnetAccount
    );
    await expect(authProvider.createLink(did)).rejects.toThrow();
  });
});

describe("authenticate", () => {
  test("Telos Testnet", async () => {
    const authProvider = new EosioAuthProvider(
      telosTestnetProvider,
      telosTestnetAccount
    );
    await expect(authProvider.authenticate("msg")).resolves.toMatchSnapshot();
  });

  test("Jungle", async () => {
    const authProvider = new EosioAuthProvider(jungleProvider, jungleAccount);
    await expect(authProvider.authenticate("msg")).resolves.toMatchSnapshot();
  });
});
