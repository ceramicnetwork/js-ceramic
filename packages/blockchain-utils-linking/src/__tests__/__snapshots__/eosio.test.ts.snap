// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accountId 1`] = `
AccountID {
  "address": "testuser1111",
  "chainId": ChainID {
    "namespace": "eosio",
    "reference": "1eaa0824707c8c16bd25145493bf062a",
  },
}
`;

exports[`authenticate Jungle 1`] = `"0xda6a3b61f0849251071ebceb94c874a3517772cc2a865c8c5e078c2e6646a0c2"`;

exports[`authenticate Telos Testnet 1`] = `"0x521674ea7031af056c7fcf69e2c3ff6f4bea888158941a7b6dc789d0fe44cf53"`;

exports[`createLink generate proof on jungle testnet 1`] = `
Object {
  "account": "idx3idctest1@eosio:2a02a0053e5a8cf73a56ba0fda11e4d9",
  "message": "Create a new account link to your identity.

did:3:bafysdfwefwe 
Timestamp: 666",
  "signature": "SIG_K1_KUxrjuxkgUuXheBQaSAeR5Ahud2LnYTf5aTtZm4qLxXQPyYVU8XAxoLPkukxHLMCPouUbFwfvYkiiuHiz6RU1ia9gT9m92",
  "timestamp": 666,
  "type": "eosio",
  "version": 1,
}
`;

exports[`createLink generate proof on telos testnet 1`] = `
Object {
  "account": "testuser1111@eosio:1eaa0824707c8c16bd25145493bf062a",
  "message": "Create a new account link to your identity.

did:3:bafysdfwefwe 
Timestamp: 666",
  "signature": "SIG_K1_Jy6okKYm2SCnv8e1qb7mUTfygmAUFG3bW9NhMMsRTUXgn92x4n67csgnjh28Q2ex2Eadf9fTCCBiFV3EA3NmLTyRZoj8sm",
  "timestamp": 666,
  "type": "eosio",
  "version": 1,
}
`;
